internal/flowcontrol/writebuffer.go

```diff
@@ -34,29 +32,32 @@ type ResourceSliceWriteBuffer struct {
 
 	// queue items are per-slice.
 	// the state map collects multiple updates per slice to be dispatched by next queue item.
-	mut   sync.Mutex
-	state map[types.NamespacedName][]*resourceSliceStatusUpdate
-	queue workqueue.RateLimitingInterface
+	mut           sync.Mutex
+	state         map[types.NamespacedName][]*resourceSliceStatusUpdate
+	insertionTime map[types.NamespacedName]time.Time
+	queue         workqueue.RateLimitingInterface
 }
 
-func NewResourceSliceWriteBufferForManager(mgr ctrl.Manager, batchInterval time.Duration, burst int) *ResourceSliceWriteBuffer {
-	r := NewResourceSliceWriteBuffer(mgr.GetClient(), batchInterval, burst)
+func NewResourceSliceWriteBufferForManager(mgr ctrl.Manager) *ResourceSliceWriteBuffer {
+	r := NewResourceSliceWriteBuffer(mgr.GetClient())
 	mgr.Add(r)
 	return r
 }
 
-func NewResourceSliceWriteBuffer(cli client.Client, batchInterval time.Duration, burst int) *ResourceSliceWriteBuffer {
+func NewResourceSliceWriteBuffer(cli client.Client) *ResourceSliceWriteBuffer {
 	return &ResourceSliceWriteBuffer{
-		client: cli,
-		state:  make(map[types.NamespacedName][]*resourceSliceStatusUpdate),
+		client:        cli,
+		state:         make(map[types.NamespacedName][]*resourceSliceStatusUpdate),
+		insertionTime: make(map[types.NamespacedName]time.Time),
 		queue: workqueue.NewRateLimitingQueueWithConfig(
-			newRateLimiter(batchInterval, burst),
+			workqueue.NewItemExponentialFailureRateLimiter(time.Millisecond*100, 8*time.Second),
```

Comment:
Since like this is deprecated in favor of https://pkg.go.dev/k8s.io/client-go/util/workqueue#NewTypedItemExponentialFailureRateLimiter in any case, are the base and max delays just heuristics or based on something?

Reply:
We're on an older version of controller-runtime that doesn't have generics yet 😢 . Created #274

Reply:
Oh I just pulled these values out of the air while playing with some simple load tests. I don't think the timing is particularly important but maybe we should expose them as flags just in case?

Reply:
I'll leave it up to you, would be nice to have the ability to overwrite them in prod, but I wouldn't mind leaving it to a follow up.